NetworkSession: add support for discarding repeated packets before th…

dktapps opened the pull request:
…ey're decoded

this dramatically reduces the server workload dealing with spammy packets like right-click interactions trigger. It also solves the problem of players getting kicked for right-clicking for too long.

<!-- Summarize your PR here. Keep it short and simple. -->
<!-- Explain existing problems or why this pull request is necessary -->

### Related issues & PRs
Fixes #5728 in local testing

## Changes
### API changes
No public API changes.

The network system has some new functions supporting this feature, but these are not part of the versioned API (as with all `pocketmine\network\mcpe` stuff).

### Behavioural changes
The server will now discard `InventoryTransactionPacket` when repeated.
This may result in a performance improvement in some servers.

## Follow-up
Maybe filter out more spammy c2s packets, like `AnimatePacket`?

## Tests
Locally tested with the "Improved input response" toggle in the client set to ON.

## Caveats
I'm not sure if doing this might have unintended side effects. So far as I know, there is no obvious reason to send the same transaction multiple times, but I may have overlooked something. Please test thoroughly and feedback.

dktapps committed:
NetworkSession: add support for discarding repeated packets before they're decoded

this dramatically reduces the server workload dealing with spammy packets like right-click interactions trigger.
It also solves the problem of players getting kicked for right-clicking for too long.

dktapps committed:
get rid of the old hack

kostamax27 commented:
https://github.com/user-attachments/assets/694eab42-c1f9-47d9-9b01-65f28eaa71e6

dktapps commented:
Not too sure how to resolve that issue. I could have it expire the filter after 1 tick, but it might still result in legit actions disappearing in certain conditions.

dktapps committed:
Merge branch 'minor-next' into repeated-packet-filter

dktapps committed:
More conservative filtering
this now only drops consecutive packets with identical buffers, and only when instructed to.
This is good enough to fix both right- and left-click spam bugs as of 1.21.100.

dktapps committed:
maybe don't make a mess?

dktapps committed:
...

dktapps committed:
... not my finest work

dktapps committed:
...

dktapps commented:
The latest commit of the PR should work better for your case @kostamax27, it's now more conservative than before

dktapps committed:
Merge branch 'minor-next' into repeated-packet-filter
